partiql-eval/src/main/java/org/partiql/eval/Statement.java

```diff
@@ -0,0 +1,20 @@
+package org.partiql.eval;
+
+import org.jetbrains.annotations.NotNull;
+import org.partiql.spi.value.Datum;
+
+/**
+ * An executable statement.
+ * <br>
+ * Developer Note: Consider `Datum execute(Parameters parameters)` for DML.
+ */
+public interface Statement {
+
+    /**
+     * Executes the statement with no parameters.
+     *
+     * @return Datum execution result.
+     */
+    @NotNull
+    public Datum execute();
+}
```

Comment:
Personally, I like some of the modeling exposed by JDBC. They have a sort of fat interface, but they don't expose a `type` -- which would actually prove quite useful for customers.
```suggestion
public interface PreparedStatement {

    /**
     * The type of the executable statement.
     */
    int type;
    
    /**
     * If a prepared statement is of this type, [executeQuery] may be invoked.
     */
    public static final int QUERY = 1;

    /**
     * Executes the query.
     *
     * @return Datum execution result.
     */
    @NotNull
    public Datum executeQuery();

    /**
     * Executes an UPDATE.
     *
     * @return UpdateDetails the details of the update
     */
    @NotNull
    public UpdateDetails executeUpdate();
}
```

Also, I don't think we need these comments regarding parameters. We've discussed this, but you can't plan without the parameters already injected, the parameters are (likely) a JDBC library feature (before sending the request to the server) and are outside of the compiler/engine.

Reply:
> but you can't plan without the parameters already injected

You can, you just need the number of parameters and optionally the types (aka the formal arguments of a function).

> (likely) a JDBC library feature

Not a JDBC thing because these parameters aren't stringified. Stringified params makes sense for JDBC because it supports multiple backends, but I think my other answers show how this adds formal arguments to a compiled statement.

ACK on the factoring of prepared statement being fat-interface like.

Reply:
Was looking into the JDBC prepared statement. We can't have an uninitialized field like in your code sample, but they may be something more interesting here. The use of executeQuery() vs executeUpdate() have to do with differing function return types; the former returns a ResultSet and the latter an int. There's also executeLargeUpdate() which returns a long. It also reads like the "executeUpdate" was added first, customers realized the limitation/bug, and the executeLargeUpdate() had to be added because a function cannot be defined with two different return types. This seems to indicate that the PreparedStatement methods don't necessarily correspond to the underlying SQL statement type, but are more about the return type.

We have two options here:

1. Datum can represent both a ResultSet and an int, long, etc; so no need to have different execution return types.
2. We introduce generics so we have `Statement<T>` with `T execute()`.

I'm working on the mental model of "A PartiQL compiler returns a statement which can be invoked with zero or more parameters". It produces _functions_ from the input environment and input string, then you can invoke that _function_.

Interestingly the JDBC statement has no notion of a statement type. I'm not against fat interfaces, but we have to keep in mind that they were a solution before generics existed and are advised against in OO design.